oak_io/src/handle.rs

```diff
@@ -59,6 +61,67 @@ impl From<Handle> for WriteHandle {
     }
 }
 
+/// Provides implementations of `Clone` and `Drop` for handle types, based on the `handle_clone` and
+/// `channel_close` ABI syscalls, respectively.
+#[cfg(feature = "linear-handles")]
+mod linear_handles {
+    use super::*;
+    use crate::OakStatus;
+    use log::trace;
+
+    /// Creates a new distinct handle using `oak_abi::clone_handle`
+    impl Clone for ReadHandle {
+        fn clone(&self) -> Self {
+            Self {
+                handle: clone_handle(self.handle),
+            }
+        }
+    }
+
+    /// Creates a new distinct handle using `oak_abi::clone_handle`
+    impl Clone for WriteHandle {
+        fn clone(&self) -> Self {
+            Self {
+                handle: clone_handle(self.handle),
+            }
+        }
+    }
+
+    /// Closes the handle using `oak_abi::channel_close`
+    impl Drop for ReadHandle {
+        fn drop(&mut self) {
+            drop_handle(self.handle);
+        }
+    }
+
+    /// Closes the handle using `oak_abi::channel_close`
+    impl Drop for WriteHandle {
+        fn drop(&mut self) {
+            drop_handle(self.handle);
+        }
+    }
+
+    fn clone_handle(handle: Handle) -> Handle {
+        let mut cloned_handle = 0;
+        let result = OakStatus::from_i32(unsafe {
+            oak_abi::handle_clone(handle, &mut cloned_handle as *mut Handle) as i32
+        })
+        .expect("handle_clone returned invalid oak status");
```

Comment:
Could you print the status too? (or does it get printed anyways?)

Reply:
It gets printed already: https://doc.rust-lang.org/std/result/enum.Result.html#method.expect

> Panics if the value is an Err, with a panic message including the passed message, **and the content of the Err**.

Reply:
Sure, but I don't know if the Err contains the status that failed to be converted :)

Reply:
Oops, `from_i32` returns an `Option`, not a `Result`. Thanks for pointing this out, will fix! 😄